storage/src/main/java/org/apache/kafka/storage/internals/epoch/LeaderEpochFileCache.java

```diff
@@ -148,6 +151,16 @@ private List<EpochEntry> removeWhileMatching(Iterator<Map.Entry<Integer, EpochEn
         return removedEpochs;
     }
 
+    public LeaderEpochFileCache writeTo(LeaderEpochCheckpoint leaderEpochCheckpoint) {
+        lock.writeLock().lock();
```

Comment:
You don't need write locks here since you are flushing to an entirely new checkpoint. You need read lock on existing cache. (similar to flush() method)

Reply:
Nice catch!